### PR TITLE
[PyTorch] Implement GQA based on fused qkv and fix layernorm for key and value

### PR DESCRIPTION
**Pull request by Marks101**
This PR implements group query attention based on a fused projection for query, key and value. This fixes issue #392 in an alternative way to PR #378.

This implementation is similar to the one in Megatron-LM. The advantage is that it generalises better to the case `num_gqa_group==num_attention_heads`. For example, the parameter naming in the state dict no longer changes when GQA is enabled or disabled. It also does not require an additional linear projection for queries, and by using `LayerNormLinear` instead of explicit normalisation, kernels should be fused correctly.

This implementation invalidates checkpoints with GQA. But considering that the current implementation suffers from issue #392, this should be OK. The notations for the dimensions get a bit messy, I would be really open to ideas for improvement ;-)

### CODE REVIEW

**Comment by cyanguwa** [comment]
Thanks @Marks101 for submitting this PR.

The `hidden_size + 2 * self.hidden_size_kv` is a neat trick but we didn't go with it in the initial implementation (https://github.com/NVIDIA/TransformerEngine/pull/338#issuecomment-1656476505) because `hidden_size + 2 * self.hidden_size_kv` might not divide well by 3 when doing `ColumnParallel` or splitting the parameters into q, k, v tensors. 

The fusion of layernorm and linear is a good idea, but if the statement in the above paragraph is valid and we'll be doing linears separately for q and kv, we're facing either `self.layernorm_query` + `self.layernorm_key_value` as our fused option, or `self.norm` + `self.query` + `self.key_value` as our unfused option. Between two instances of layernorm and one instance of layernorm but unfused, I'm not sure how this is going to impact the code performance wise.

Any ideas are welcome. @adhocyyf #378

Thanks.

**Comment by wangye01inf** [comment]
Hi @cyanguwa @Marks101, here are some of my thoughts：

The fusion of input norm and qkv is interesting，but may encounter with problems when `fuse_qkv_params=False` in `Linear` module, which will `assert self.out_features % len(parameters_split) == 0`

The idea of `self.layernorm_query + self.layernorm_key_value` seems strange because we will have duplicate params of layernorm from `layernorm_query` and `layernorm_key_value`

I also tried an alternative way to fix this，which create `layernorm_query` with `return_layernorm_output=True`，and reuse the `layernorm_output` in forward. But it report RuntimeError in `layernorm_query` forward stage, which i post the picture below.

I suggest that we can fix this issue firstly，then create new Issues about performance or `return_layernorm_output`

image

Thanks ;-)

**Comment by adhocyyf** [comment]
Hi @cyanguwa @Marks101, here are some of my thoughts：

The fusion of input norm and qkv is interesting，but may encounter with problems when `fuse_qkv_params=False` in `Linear` module, which will `assert self.out_features % len(parameters_split) == 0`

The idea of `self.layernorm_query + self.layernorm_key_value` seems strange because we will have duplicate params of layernorm from `layernorm_query` and `layernorm_key_value`

I also tried an alternative way to fix this，which create `layernorm_query` with `return_layernorm_output=True`，and reuse the `layernorm_output` in forward. But it report RuntimeError in `layernorm_query` forward stage, which i post the picture below.

I suggest that we can fix this issue firstly，then create new Issues about performance or `return_layernorm_output`

image

Thanks ;-)

**Comment by Marks101** [comment]
Hi @cyanguwa  and @adhocyyf thank you for the quick replies and all your thoughts about this topic.

> because hidden_size + 2 * self.hidden_size_kv might not divide well by 3 when doing ColumnParallel or splitting the parameters into q, k, v tensors.

Okay, the issue lies in `fuse_qkv_params=False`. I did not see that before. Thanks for the clarification!

I started to do a few benchmarks for `MultiheadAttention` in order to evaluate the performance of the different implementations. I added the benchmark script at the bottom. The benchmark was done on a H100 DGX with NCCL 2.18.3, CUDA 12.1, torch 2.1.0 (dev) and flash-attn 2.0.4. The command I used is `torchrun --nproc-per-node TP_SIZE benchmark_multiheadattention.py --num_gqa_groups 8`. The overall setting is typical for a 40B model. The timings I get are:

| branch         | commit | timing TP_SIZE=1 [ms] | timing TP_SIZE=8 [ms] |
| -----------  | -------- | ---------------------- | ----------------------- |
| main branch | 3b7b7c68| 17.2| 4.53        |
| #378             | cc81e01f | 17.2 | 4.53        |
| this PR          | 6f206fac | 17.2 | 3.71         |

We see that, at least in these tests, having a separate normalization in #378 is no performance issue, but having separate projections for query and (key, value) has a massive implication when tensor parallelism is active. In this benchmark the fused qkv implementation is around 18% faster. I tested it in an actual training is well, there the step time differs by around 9%.

Could you maybe try to replicate these numbers? Perhaps this is just something special about my setup.

My feeling is that GQA got very popular, for this reason good performance numbers for this architicture matter for many people.

<details>
  <summary>Benchmark script</summary>
  <p>

    import argparse
    from typing import List, Optional
    import time
    import torch
    import transformer_engine as te
    
    
    def speedometer(
        module: torch.nn.Module,
        args: List[torch.Tensor],
        output_grad: torch.Tensor,
        fp8_autocast_kwargs: Optional[dict] = None,
        timing_iters: int = 500,
        warmup_iters: int = 100,
    ) -> None:
        """Measure average run time for a PyTorch module
    
        Performs forward and backward passes.
        """
        if fp8_autocast_kwargs is None:
            fp8_autocast_kwargs = {"enabled": False}
    
        # Warmup runs
        for _ in range(warmup_iters):
            with te.pytorch.fp8_autocast(**fp8_autocast_kwargs):
                output = module(*args)
            output.backward(output_grad)
    
        # Timing runs
        torch.distributed.barrier()
        torch.cuda.synchronize()
        start = time.time()
    
        for _ in range(timing_iters):
            with te.pytorch.fp8_autocast(**fp8_autocast_kwargs):
                output = module(*args)
            output.backward(output_grad)
    
        torch.distributed.barrier()
        torch.cuda.synchronize()
        elapsed = time.time() - start
    
        if torch.distributed.get_rank() == 0:
            print(f"Rank 0 mean time: {elapsed/timing_iters} s")
    
    
    if __name__ == "__main__":
        parser = argparse.ArgumentParser()
        parser.add_argument("--hidden_size", default=8192, type=int)
        parser.add_argument("--num_attention_heads", default=64, type=int)
        parser.add_argument("--num_gqa_groups", default=64, type=int)
        parser.add_argument("--sequence_length", default=2048, type=int)
        parser.add_argument("--batch_size", default=4, type=int)
        parser.add_argument("--fp8", action="store_true")
        args = parser.parse_args()
    
        torch.distributed.init_process_group(backend="nccl")
        tp_group = torch.distributed.new_group(
            ranks=range(torch.distributed.get_world_size())
        )
        torch.cuda.set_device(torch.distributed.get_rank())
    
        model = te.pytorch.MultiheadAttention(
            args.hidden_size,
            args.num_attention_heads,
            num_gqa_groups=args.num_gqa_groups,
            fuse_qkv_params=True,
            tp_group=tp_group,
            set_parallel_mode=True,
            params_dtype=torch.bfloat16,
        ).cuda()
    
        inp_shape = (
            args.sequence_length,
            args.batch_size,
            args.hidden_size,
        )
    
        inp = torch.randn(
            inp_shape, dtype=torch.bfloat16, requires_grad=True, device="cuda"
        )
    
        output_grad = torch.randn(inp_shape, dtype=torch.bfloat16, device="cuda")
    
        recipe = te.common.recipe.DelayedScaling(fp8_format=te.common.recipe.Format.HYBRID)
        fp_autocast_kwargs = {"enabled": True, "fp8_recipe": recipe}
        speedometer(
            model,
            [inp],
            output_grad,
            fp8_autocast_kwargs=fp_autocast_kwargs if args.fp8 else None,
        )
  </p>
</details>

**Comment by cyanguwa** [comment]
@Marks101 Yes, the parameter split is a problem if we use `hidden_size + 2 * hidden_size_kv`. But also, `ColumnParallel` might have some issue when TP is present. Currently we only do 'even split' for `parameters_split` (in `Linear` and `LayerNormLinear`), and for `ColumnParallel` (when `parallel_mode = column` in MHA). When it's not divisible by 3, we would be splitting q,k,v the wrong way, and putting different workloads across the TP GPUs as well. Both should be fixable though with a bit of engineering effort and testing, but I guess we just don't have it yet as of now.

Thanks for the performance benchmarking. I agree we should squeeze out every bit of performance we can - maybe we should really evaluate if fusion is even possible here. @ptrendx 

@adhocyyf I agree we shouldn't duplicate the layernorm params by using `self.layernorm_query` and `self.layernorm_key_value`.

Regarding the `return_layernorm_output` error, it looks like the code has gone through the FP8 path? Maybe try running it with FP16/BF16?

Thanks!

**Comment by Marks101** [comment]
@cyanguwa Thanks for the clarifications. But as long as `fuse_qkv_params==True` and `num_gqa_groups % tp_size == 0`, there should be no problem in my fused implementation, or am I missing something? The assumpiton `num_gqa_groups % tp_size == 0` seems very natural to me for GQA. The line `self.num_gqa_groups_per_partition = int(self.num_gqa_groups // tp_size))` already assumes it. This is actually not asserted in the current main branch, here `mixed_x_layer.view(*new_tensor_shape)` fails if this assumption is not met.

All in all, I see two options

- Going with #378: in this case it will be hard to squeeze out the full performance without changing the checkpoint layout at a later point of time.
- Going with the fused implementation in my PR: in this case it should be asserted that fuse_qkv_params==True if GQA is active. Later on, it might be possible to add uneven parameter splits to the `Linear` to remove this restriction again.

What are your thoughts?

**Comment by cyanguwa** [comment]
Hi @Marks101, you're right, I think there was some misconception there with my comment. For `ColumnParallel` to work (i.e. load-balanced across TP), we need `hidden_size + 2 * hidden_size_kv` to be divisible by `tp_size`, not by 3. And our current assumptions, `num_gqa_groups % tp_size == 0` and `num_attention_heads % self.num_gqa_groups == 0`, should suffice for that! Let's continue to work under these assumptions and add support for when they're not true later. 

On the other hand, for `parameters_split` to work, we need `hidden_size + 2 * hidden_size_kv` to be divisible by 3 (in the self attention case, not a problem in cross attention). You're right that we should provide the high performance path, i.e. your fused implementation, but can we make it work for both `fuse_qkv_params == True` and `fuse_qkv_params == False`? For example, can we add a dictionary support for `parameters_split`: `Union[Tuple[str, ...], Dict[str, int]]` so that if the input is `("query_", "key_", "value_")` it assumes an even split and if it's `{"query_": hidden_size, "key_": hidden_size_kv, "value_": hidden_size_kv}` it uses `hidden_size` and `hidden_size_kv` to split? I think the other change we need to make here is probably just the indexing [here ](https://github.com/NVIDIA/TransformerEngine/blob/a5dbf1e29f31a053107aec80731511e73c20b768/transformer_engine/pytorch/module/layernorm_linear.py#L726)and [here](https://github.com/NVIDIA/TransformerEngine/blob/a5dbf1e29f31a053107aec80731511e73c20b768/transformer_engine/pytorch/module/layernorm_linear.py#L738). 

Thanks!

**Comment by Marks101** [comment]
Hi @cyanguwa, I really like your suggestion and implemented it. Please have a look

**Comment by cyanguwa** [comment]
Thanks! It looks good to me. I'll run some CI tests.

Thanks,
Charlene

**Comment by cyanguwa** [comment]
@Marks101 I can't seem to push to your branch from the command line. I'm working in PR #434 for now.

**Comment by cyanguwa** [comment]
#434 is merged! Thanks!

**Comment by Marks101** [comment]
Thank you so much @cyanguwa for putting the effort into this to merge it :D
